recipes/magnum/all/cmake/conan-bugfix-global-target.cmake

```diff
@@ -0,0 +1,24 @@
+# If using 'cmake_find_package', all the components are being added to the global
```

Comment:
Can't we work around this by giving the "global" target another name?
e.g.
```cmake
self.cpp_info.filenames["cmake_find_package_multi"] = "Magnum"
self.cpp_info.set_property("cmake_target_name", "MagnumGlobal")
self.cpp_info.components["libmagnum"].names["cmake_find_package_multi"] = "Magnum"
```

Reply:
The problem with `cmake_find_package` is not about the global-target name, it is a bug with the generator itself. 

There are these lines in the generated `FindMagnum.cmake` file to create the global target:

```cmake
if(NOT ${CMAKE_VERSION} VERSION_LESS "3.0")
    if(NOT TARGET Magnum::Magnum)
        add_library(Magnum::Magnum INTERFACE IMPORTED)
    endif()
    set_property(TARGET Magnum::Magnum APPEND PROPERTY
                 INTERFACE_LINK_LIBRARIES "${Magnum_COMPONENTS}")
endif()
```

That `set_property(TARGET Magnum::Magnum APPEND PROPERTY "${Magnum_COMPONENTS}")` will append a list with all the components. As a consequence, consumer libraries will try to link with all their libraries, but the linker won't be able to find them because **these libraries are in a different folder** (not in `/lib`)... the required _link directories_ are not added to the `Magnum::Magnum` target, and it fails:

```
/usr/bin/ld: cannot find -lAnyAudioImporter
/usr/bin/ld: cannot find -lAnyImageConverter
/usr/bin/ld: cannot find -lAnyImageImporter
/usr/bin/ld: cannot find -lAnySceneConverter
/usr/bin/ld: cannot find -lAnySceneImporter
/usr/bin/ld: cannot find -lMagnumFont
/usr/bin/ld: cannot find -lMagnumFontConverter
/usr/bin/ld: cannot find -lObjImporter
/usr/bin/ld: cannot find -lTgaImporter
/usr/bin/ld: cannot find -lTgaImageConverter
/usr/bin/ld: cannot find -lWavAudioImporter
```

The `Magnum::Magnum` target is broken (overridden or not).

---

There are two possible fixes:
 * Add the required folders to find all the libraries
 * Override the `INTERFACE_LINK_LIBRARIES` so it links just the targets it has to link (this is the current behavior of `cmake_find_package_multi`)
 
I've opted for the second one, which will create the expected targets and is also aligned with the current behavior of some other generators.

Reply:
Shouldn't this bug/problem be fixed in conan then?
Would the problem disappear if `"Magnum_COMPONENTS` is a list of targets?

Reply:
it's very unlikely bug will be ever fixed in `cmake_find_package`.
first, because this generator will die very soon, so it's pointless to spend time on that for conan developers.
second, even if there is a bug, the fix will be considered "breaking". even if behavior is incorrect, there might be users already relying on this incorrect behavior.
the only thing that could be done is new property (opt-out) to override the behavior of the generator, but for the first reason, it's unlikely to happen anyway.

Reply:
Agree. But most of the effort is focused on new generators (`CMakeDeps`), so it is probably that `cmake_find_package` won't get enough attention, and the fix could only be in the wrong direction: adding link directories. This is the only way to not break existing consumers that are using the global target to link with everything... although the recipe author was intending to limit the components exposed in the global target.

I think it is not only required that `Magnum_COMPONENTS` is a list of targets (IIRC it already is). It is required to use some `target_link_libraries(Magnum::Magnum INTERFACE ${Magnum_COMPONENTS})` (not tested) so everything is propagated from those targets.

Any of these changes will differ even more from `cmake_find_package_multi` which is doing it right.

Reply:
I would add #FIXME tag, so we can filter it in the future and remove once fixed by Conan

Reply:
The file is "conan bugfix" but a `FIXME` is easier to find with search